azurerm/internal/services/compute/linux_virtual_machine_resource.go

```diff
@@ -1030,7 +1101,76 @@ func resourceLinuxVirtualMachineUpdate(d *schema.ResourceData, meta interface{})
 		} else {
-			return fmt.Errorf("Once a customer-managed key is used, you can’t change the selection back to a platform-managed key")
+			return fmt.Errorf("once a customer-managed key is used, you can’t change the selection back to a platform-managed key")
+		}
+	}
+
+	if features.VMDataDiskBeta() && d.HasChanges("data_disks.0.local", "data_disks.0.existing") {
+		shouldUpdate = true
+		dataDisks, err := expandVirtualMachineDataDisks(d, meta)
+		if err != nil {
+			return err
+		}
+
+		// Do we have disks to resize or change encryption set?
+		for _, v := range dataDisksForEncrypt {
+			diskName, ok := v["name"].(string)
+			if !ok {
+				return fmt.Errorf("could not resize data disk, empty value for `name`")
+			}
+			diskEncryptionSetID, ok := v["disk_encryption_set_id"].(string)
+			if !ok {
+				return fmt.Errorf("failed to read new disk Encryption Eet ID for Data Disk %q (resource group %q)", diskName, id.ResourceGroup)
+			}
+			diskUpdate := compute.DiskUpdate{
+				DiskUpdateProperties: &compute.DiskUpdateProperties{
+					Encryption: &compute.Encryption{
+						DiskEncryptionSetID: utils.String(diskEncryptionSetID),
+					},
+				},
+			}
+
+			encryptionUpdateFuture, err := disksClient.Update(ctx, id.ResourceGroup, diskName, diskUpdate)
+			if err != nil {
+				return fmt.Errorf("failed resizing Data Disk %q for Virtual Machine %q (resource group %q): %+v", diskName, id.Name, id.ResourceGroup, err)
+			}
+			if err = encryptionUpdateFuture.WaitForCompletionRef(ctx, client.Client); err != nil {
+				return fmt.Errorf("failed waiting for resize of Data Disk %q for Virtual Machine %q (resource group %q): %+v", diskName, id.Name, id.ResourceGroup, err)
+			}
+		}
+
+		for _, v := range dataDisksForGrow {
+			diskName, ok := v["name"].(string)
+			if !ok {
+				return fmt.Errorf("could not resize data disk, empty value for `name`")
+			}
+			diskSizeGB, ok := v["disk_size_gb"].(int)
+			if !ok {
+				return fmt.Errorf("failed to read new disk size for Data Disk %q (resource group %q)", diskName, id.ResourceGroup)
+			}
+
+			diskUpdate := compute.DiskUpdate{
+				DiskUpdateProperties: &compute.DiskUpdateProperties{
+					DiskSizeGB: utils.Int32(int32(diskSizeGB)),
+				},
+			}
+
+			resizeFuture, err := disksClient.Update(ctx, id.ResourceGroup, diskName, diskUpdate)
+			if err != nil {
+				return fmt.Errorf("failed resizing Data Disk %q for Virtual Machine %q (resource group %q): %+v", diskName, id.Name, id.ResourceGroup, err)
+			}
+			if err = resizeFuture.WaitForCompletionRef(ctx, client.Client); err != nil {
+				return fmt.Errorf("failed waiting for resize of Data Disk %q for Virtual Machine %q (resource group %q): %+v", diskName, id.Name, id.ResourceGroup, err)
+			}
+		}
+
+		// now disks are resized, we can put the new list into the VMUpdate
+		if update.VirtualMachineProperties.StorageProfile == nil {
+			update.VirtualMachineProperties.StorageProfile = &compute.StorageProfile{
+				DataDisks: dataDisks,
```

Comment:
this should never be nil, since the OS Disk is captured within it - so I think this'd be invalid?

Reply:
Only if `os_disk` has a change also, since it's a Patch the StorageProfile will be nil here if only the Data Disk(s) have change?